Storage/Storage/CoreData/CoreDataManager.swift

```diff
@@ -172,6 +173,32 @@ public final class CoreDataManager: StorageManagerType {
         })
     }
 
+    /// Execute the given `operation` with a background context, save the changes, and return the result.
+    ///
+    /// This function _does not block_ its running thread. The operation is executed in background and its return value
+    /// is passed onto the `completion` closure which is executed on the given `queue`.
+    ///
+    /// - Parameters:
+    ///   - operation: A closure which uses the given `StorageType` to make data changes in background.
+    ///   - completion: A closure which is called with the `operation`'s execution result,
+    ///   which is either an error thrown by the `operation` or the return value of the `operation`.
+    ///   - queue: A queue on which to execute the completion closure.
+    public func performAndSave<T>(_ operation: @escaping (StorageType) throws -> T,
+                                  completion: @escaping (Result<T, Error>) -> Void,
+                                  on queue: DispatchQueue) {
+        let derivedStorage = writerDerivedStorage
+        writerQueue.addOperation(AsyncBlockOperation { done in
+            derivedStorage.perform {
+                let result = Result(catching: { try operation(derivedStorage) })
+                if case .success = result {
+                    derivedStorage.saveIfNeeded()
+                }
```

Comment:
Would it be helpful to log the error here if happens? Or wouldn't be necessary since we move the responsibility elsewhere and we just propagate it? 🤔 

```suggestion
                if case .success = result {
                    derivedStorage.saveIfNeeded()
                }
                if case .failure(let error) = result {
                    DDLogError( ... 
                }
```

Reply:
I believe logging is up to the developer writing the operation closure. In my case with `markOrderAsPaidLocally`, the failure is intentional and I handle it in the completion closure. So, logging in `CoreDataManager` is not necessary.